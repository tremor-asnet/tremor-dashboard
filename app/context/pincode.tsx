"use client";

import {
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

interface IPinCodeContext {
  isConfirmPinCode: boolean;
  pinCode?: number;
  isShowPinCodeModal: boolean;
  setPinCode: (code: number) => void;
  showPinCodeModal: () => void;
  hidePinCodeModal: () => void;
  confirmPinCode: (code: number) => boolean;
}

const initialPinCodeContext: IPinCodeContext = {
  isConfirmPinCode: false,
  confirmPinCode: () => false,
  hidePinCodeModal: () => {},
  setPinCode: () => false,
  showPinCodeModal: () => {},
  isShowPinCodeModal: false,
};

const PinCodeContext = createContext(initialPinCodeContext);

const PinCodeProvider = ({ children }: { children: ReactNode }) => {
  const [isConfirmPinCode, setIsConfirmPinCode] = useState(false);
  const [isShowPinCodeModal, setIsShowPinCodeModal] = useState(false);
  const [pinCode, setPinCode] = useState<number>();

  useEffect(() => {
    setIsShowPinCodeModal(!pinCode);
  }, [pinCode]);

  const handleShowPinCodeModal = useCallback(
    () => setIsShowPinCodeModal(true),
    [],
  );

  const handleHidePinCodeModal = useCallback(
    () => setIsShowPinCodeModal(false),
    [],
  );

  const handleConfirmPinCode = useCallback(
    (code: number) => {
      const isMatch = code === pinCode;
      setIsConfirmPinCode(isMatch);
      isMatch ? handleHidePinCodeModal() : handleShowPinCodeModal();
      return isMatch;
    },
    [handleHidePinCodeModal, handleShowPinCodeModal, pinCode],
  );

  const pinCodeContextValue: IPinCodeContext = {
    isConfirmPinCode,
    confirmPinCode: handleConfirmPinCode,
    pinCode,
    isShowPinCodeModal,
    setPinCode,
    hidePinCodeModal: handleHidePinCodeModal,
    showPinCodeModal: handleShowPinCodeModal,
  };

  return (
    <PinCodeContext.Provider value={pinCodeContextValue}>
      {children}
    </PinCodeContext.Provider>
  );
};

const usePinCode = () => {
  const context = useContext(PinCodeContext);

  if (!context) {
    throw new Error("usePinCode hooks should using inside PinCodeProvider!");
  }

  return context;
};

export { PinCodeProvider, usePinCode };
