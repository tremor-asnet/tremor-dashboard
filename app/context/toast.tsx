"use client";

import React, { createContext, useState } from "react";

// Components
import { FaCheckCircle } from "react-icons/fa";
import { TbExclamationMark } from "react-icons/tb";
import { RxCross2 } from "react-icons/rx";
import Toast from "@/ui/components/Toast/Toast";

export enum TOAST_TYPE {
  SUCCESS = "success",
  WARNING = "warning",
  ERROR = "error",
}
interface ToastProps {
  type?: TOAST_TYPE;
  message?: string;
  icon?: JSX.Element;
}

type ToastState = ToastProps & { isOpen: boolean };

interface ToastContextValue {
  openToast: (toastProps: ToastProps) => void;
  closeToast: () => void;
}

export const buildToastRenderer = ({ type, icon, message }: ToastProps) => {
  switch (type) {
    case TOAST_TYPE.WARNING:
      return {
        icon: icon || TbExclamationMark,
        message: message ?? "Warning!",
        color: "yellow",
      };

    case TOAST_TYPE.ERROR:
      return {
        icon: icon || RxCross2,
        message: message ?? "Error!",
        color: "red",
      };

    case TOAST_TYPE.SUCCESS:
    default:
      return {
        icon: icon || FaCheckCircle,
        message: message ?? "Success!",
        color: "green",
      };
  }
};

const ToastContext = createContext<ToastContextValue>({
  openToast: () => {},
  closeToast: () => {},
});

const ToastProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [toast, setToast] = useState<ToastState>({
    isOpen: false,
    type: TOAST_TYPE.SUCCESS,
  });

  const closeToast = () => {
    setToast({
      ...toast,
      isOpen: false,
    });
  };

  const openToast = ({
    type = TOAST_TYPE.SUCCESS,
    message,
    icon,
  }: ToastProps) => {
    setToast({ isOpen: true, type, message, icon });

    setTimeout(() => {
      closeToast();
    }, 3000);
  };

  const toastRenderer = buildToastRenderer(toast);
  return (
    <ToastContext.Provider
      value={{
        openToast,
        closeToast,
      }}>
      {children}
      {toast.isOpen && (
        <Toast
          Icon={toastRenderer.icon}
          message={toastRenderer.message}
          color={toastRenderer.color}
          onClose={closeToast}
        />
      )}
    </ToastContext.Provider>
  );
};

export { ToastProvider, ToastContext };
