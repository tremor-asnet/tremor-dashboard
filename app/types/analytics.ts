export interface IAnalyticsInfo {
  id: string;
  photo: string;
  name: string;
  description: string;
  price: number;
  location: string;
}

export enum CHART_TYPE {
  SALE = "sale",
  PERFORMANCE = "performance",
}

export type LINE_CHART_DATA = {
  date: string;
  "Desktop apps"?: number;
  "Mobile apps"?: number;
};

export type LINE_CHART = {
  id: string;
  display: string;
  descValue: string;
  desc: string;
  modified: string;
  data: LINE_CHART_DATA[];
};

type TFixedProps = {
  eventType: "dot" | "category" | "bar" | "slice" | "bubble";
  categoryClicked: string;
};

type TBaseEventProps = TFixedProps & {
  [key: string]: number | string;
};

export type TEventProps = TBaseEventProps | null | undefined;
