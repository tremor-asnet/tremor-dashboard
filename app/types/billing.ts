export interface Invoice {
  date: string;
  code: string;
  price: number;
}

export interface Transaction {
  id: number;
  createdAt: string;
  service: string;
  amount: number;
  type: number;
  status: number;
}
export interface BillingCardData {
  id: number;
  type: string;
  expire: string;
  cardNumber: string;
  holderFullName: string;
  cardLast4Digit: string;
}
