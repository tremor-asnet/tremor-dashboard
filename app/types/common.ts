export interface Step {
  index: number;
  content: string;
}

export interface IMedia {
  image: string;
}

export interface CdnResponse {
  data: {
    width: number;
    height: number;
    image: ImageInfo;
  };
}

export interface ImageInfo {
  filename: string;
  url: string;
}

export interface SortItem {
  key: string;
  direction: string;
}

export type TSearchParams = {
  query?: string;
  isAvailable?: string;
  status?: string;
  page?: number;
  sortBy?: string;
  orderBy?: string;
};
