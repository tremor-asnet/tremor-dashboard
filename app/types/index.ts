export * from "./user";
export * from "./breadcrumb";
export * from "./tab";
export * from "./project";
export * from "./analytics";
export * from "./sales";
export * from "./profile";
export * from "./table-list";
export * from "./orders";
