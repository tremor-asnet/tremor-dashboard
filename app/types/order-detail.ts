export type ProductDetail = {
  id: number;
  name: string;
  count: number;
  price: number;
  url: string;
};

export type TOrderDetail = {
  id: number;
  order_code: string;
  order_deliver_price: number;
  order_tax: string;
  created_at: string;
  products: ProductDetail[];
  track_order_info: {
    generate_order_id: string;
    generate_order_at: string;
    transmited_to_courier_at: string;
    delivered_at: string;
  };
  billing_info: {
    last4_bank: string;
    owner_name: string;
    company_name: string;
    email: string;
    vat: string;
  };
};
