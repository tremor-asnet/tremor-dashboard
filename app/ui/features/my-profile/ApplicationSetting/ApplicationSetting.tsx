"use client";
import { Switch, Text } from "@tremor/react";

// Constants
import {
  APPLICATION_SETTING_FIELDS,
  APPLICATION_SETTING_DATA,
} from "@/constants";

// Types
import { PlatformSetting, ApplicationSettingData } from "@/types";

// Contexts
import { useContext } from "react";

// Themes
import { ThemeContext } from "@/context/theme";

interface ApplicationSettingProps {
  applicationSettingFields: PlatformSetting[];
  applicationSettingData: ApplicationSettingData;
}

const ApplicationSetting = ({
  applicationSettingFields = APPLICATION_SETTING_FIELDS,
  applicationSettingData = APPLICATION_SETTING_DATA,
}: ApplicationSettingProps) => {
  const { isDarkTheme } = useContext(ThemeContext);
  const color = isDarkTheme ? "white" : "zinc";

  return (
    <>
      <Text className="text-xs dark:text-dark-romance leading-tight opacity-100 uppercase no-underline text-[#7b809a] font-bold m-0 pt-4">
        Application
      </Text>
      {applicationSettingFields.map(({ label, field }: PlatformSetting) => (
        <div
          data-testid="application-setting"
          className="flex items-center space-x-3 mt-1 py-3"
          key={label}>
          <Switch
            tabIndex={2}
            id="switch"
            name="switch"
            color={color}
            className="flex justify-center items-center"
            defaultChecked={applicationSettingData[field]}
          />
          <Text className="text-secondary font-normal dark:text-dark-romance">
            {label}
          </Text>
        </div>
      ))}
    </>
  );
};

export default ApplicationSetting;
