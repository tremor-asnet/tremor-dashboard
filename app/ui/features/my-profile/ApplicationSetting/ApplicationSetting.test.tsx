import { fireEvent, render, screen } from "@testing-library/react";

// Contexts
import { ThemeContext, ThemeProvider } from "@/context/theme";

// Components
import ApplicationSetting from "./ApplicationSetting";

// Types
import { ApplicationSettingData } from "@/types";

// Constants
import {
  APPLICATION_SETTING_DATA,
  APPLICATION_SETTING_FIELDS,
} from "@/constants";

describe("ApplicationSetting component", () => {
  const props = {
    applicationSettingFields: APPLICATION_SETTING_FIELDS,
    applicationSettingData: APPLICATION_SETTING_DATA,
  };

  const propsEmptyData = {
    applicationSettingFields: [],
    applicationSettingData: {} as ApplicationSettingData,
  };

  it("Should render ApplicationSetting snapshot correctly", () => {
    const { getByTestId, container } = render(
      <ThemeProvider>
        <ThemeContext.Consumer>
          {({ isDarkTheme, toggleTheme }) => (
            <>
              <ApplicationSetting {...props} />
              <button
                data-testid="toggle-theme"
                onClick={() =>
                  toggleTheme()
                }>{`isDarkTheme to ${isDarkTheme}`}</button>
            </>
          )}
        </ThemeContext.Consumer>
      </ThemeProvider>,
    );
    const themeButton = getByTestId("toggle-theme");

    fireEvent.click(themeButton);
    expect(themeButton.textContent).toBe("isDarkTheme to true");

    fireEvent.click(themeButton);
    expect(themeButton.textContent).toBe("isDarkTheme to false");

    expect(container).toMatchSnapshot();
  });

  it("Should render ApplicationSetting in case empty data", () => {
    render(<ApplicationSetting {...propsEmptyData} />);

    expect(screen.queryByTestId("application-setting")).toBeNull();
  });
});
