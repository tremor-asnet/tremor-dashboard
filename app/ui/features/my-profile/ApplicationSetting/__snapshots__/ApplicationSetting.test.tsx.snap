// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationSetting component Should render ApplicationSetting snapshot correctly 1`] = `
<div>
  <p
    class="text-tremor-default text-xs dark:text-dark-romance leading-tight opacity-100 uppercase no-underline text-[#7b809a] font-bold m-0 pt-4"
  >
    Application
  </p>
  <div
    class="flex items-center space-x-3 mt-1 py-3"
    data-testid="application-setting"
  >
    <div
      class="flex flex-row items-center justify-start"
    >
      <div
        class="flex justify-center items-center"
        tabindex="2"
      >
        <input
          class="tremor-Switch-input absolute w-5 h-5 cursor-pointer left-0 top-0 opacity-0"
          name="switch"
          type="checkbox"
        />
        <button
          aria-checked="false"
          class="tremor-Switch-switch w-10 h-5 group relative inline-flex shrink-0 cursor-pointer items-center justify-center rounded-tremor-full focus:outline-none"
          data-headlessui-state=""
          id="switch"
          role="switch"
          tabindex="0"
          type="button"
        >
          <span
            class="tremor-Switch-sr-only sr-only"
          >
            Switch 
            off
          </span>
          <span
            aria-hidden="true"
            class="tremor-Switch-background bg-tremor-border dark:bg-dark-tremor-border pointer-events-none absolute mx-auto h-3 w-9 rounded-tremor-full transition-colors duration-100 ease-in-out"
          />
          <span
            aria-hidden="true"
            class="tremor-Switch-round translate-x-0 bg-tremor-border dark:bg-dark-tremor-border border-tremor-background dark:border-dark-tremor-background pointer-events-none absolute left-0 inline-block h-5 w-5 transform rounded-tremor-full border-2 shadow-tremor-input duration-100 ease-in-out transition"
          />
        </button>
      </div>
    </div>
    <p
      class="text-tremor-default text-secondary font-normal dark:text-dark-romance"
    >
      New launches and projects
    </p>
  </div>
  <div
    class="flex items-center space-x-3 mt-1 py-3"
    data-testid="application-setting"
  >
    <div
      class="flex flex-row items-center justify-start"
    >
      <div
        class="flex justify-center items-center"
        tabindex="2"
      >
        <input
          class="tremor-Switch-input absolute w-5 h-5 cursor-pointer left-0 top-0 opacity-0"
          name="switch"
          type="checkbox"
        />
        <button
          aria-checked="false"
          class="tremor-Switch-switch w-10 h-5 group relative inline-flex shrink-0 cursor-pointer items-center justify-center rounded-tremor-full focus:outline-none"
          data-headlessui-state=""
          id="switch"
          role="switch"
          tabindex="0"
          type="button"
        >
          <span
            class="tremor-Switch-sr-only sr-only"
          >
            Switch 
            off
          </span>
          <span
            aria-hidden="true"
            class="tremor-Switch-background bg-tremor-border dark:bg-dark-tremor-border pointer-events-none absolute mx-auto h-3 w-9 rounded-tremor-full transition-colors duration-100 ease-in-out"
          />
          <span
            aria-hidden="true"
            class="tremor-Switch-round translate-x-0 bg-tremor-border dark:bg-dark-tremor-border border-tremor-background dark:border-dark-tremor-background pointer-events-none absolute left-0 inline-block h-5 w-5 transform rounded-tremor-full border-2 shadow-tremor-input duration-100 ease-in-out transition"
          />
        </button>
      </div>
    </div>
    <p
      class="text-tremor-default text-secondary font-normal dark:text-dark-romance"
    >
      Monthly product updates
    </p>
  </div>
  <div
    class="flex items-center space-x-3 mt-1 py-3"
    data-testid="application-setting"
  >
    <div
      class="flex flex-row items-center justify-start"
    >
      <div
        class="flex justify-center items-center"
        tabindex="2"
      >
        <input
          class="tremor-Switch-input absolute w-5 h-5 cursor-pointer left-0 top-0 opacity-0"
          name="switch"
          type="checkbox"
        />
        <button
          aria-checked="false"
          class="tremor-Switch-switch w-10 h-5 group relative inline-flex shrink-0 cursor-pointer items-center justify-center rounded-tremor-full focus:outline-none"
          data-headlessui-state=""
          id="switch"
          role="switch"
          tabindex="0"
          type="button"
        >
          <span
            class="tremor-Switch-sr-only sr-only"
          >
            Switch 
            off
          </span>
          <span
            aria-hidden="true"
            class="tremor-Switch-background bg-tremor-border dark:bg-dark-tremor-border pointer-events-none absolute mx-auto h-3 w-9 rounded-tremor-full transition-colors duration-100 ease-in-out"
          />
          <span
            aria-hidden="true"
            class="tremor-Switch-round translate-x-0 bg-tremor-border dark:bg-dark-tremor-border border-tremor-background dark:border-dark-tremor-background pointer-events-none absolute left-0 inline-block h-5 w-5 transform rounded-tremor-full border-2 shadow-tremor-input duration-100 ease-in-out transition"
          />
        </button>
      </div>
    </div>
    <p
      class="text-tremor-default text-secondary font-normal dark:text-dark-romance"
    >
      Subscribe to newsletter
    </p>
  </div>
  <button
    data-testid="toggle-theme"
  >
    isDarkTheme to false
  </button>
</div>
`;
